kame/kame/v6test/v6test.1
.\"	$KAME: v6test.1,v 1.5 2000/11/09 03:04:13 itojun Exp $
.\"
.\" Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 17, 1998
.Dt V6TEST 1
.Os KAME
.\"
.Sh NAME
.Nm v6test
.Nd send IPv6 test packets
.\"
.Sh SYNOPSIS
.Nm
.Op Fl s Ar src
.Op Fl d Ar dst
.Op Fl f Ar configfile
.Op Fl i Ar interface
.Ar testname ...
.\"
.Sh DESCRIPTION
.Nm V6test
generates IPv6 packet described in
.Ar configfile .
This is usually used for debugging the behavior of kernel against
ill-formatted packets.
.Bl -tag -width Ds
.It Fl s Ar src
Source IPv6 address for the test packet.
The option lets you override the description in configuration file.
.It Fl d Ar dst
Destination IPv6 address for the test packet.
The option lets you override the description in configuration file.
.It Fl i Ar interface
Send the generated IPv6 packet to
.Ar interface .
.Xr bpf 4
encapsulation type
.Dv DLT_EN10MB
and
.Dv DLT_NULL
are supported.
.It Fl f Ar configfile
Specify alternate location to configuration file.
Default is
.Pa v6test.conf
in the current directory.
.It Ar testname
Specify a single test entry in the configuration file.
You can specify multiple tests, to send contiguous test packets.
.El
.\"
.Sh RETURN VALUES
.Nm V6test
exits with 0 on success, and non-zero on errors.
.\"
.Sh FILES
.Bl -tag -width ./v6test.conf -compact
.It Pa ./v6test.conf
The default configuration file.
.El
.\"
.Sh SEE ALSO
.Xr bpf 4 ,
.Xr tcpdump 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE Hydrangea IPv6 protocol stack kit.
.\"
.Sh BUGS
Man page for
.Xr v6test.conf 5
has to be written.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/v6test/v6test.c
/*
 * Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "common.h"
#ifdef HAVE_LIBPCAP
#include <pcap.h>
#endif 

u_char buf[65536];
char *conffile = NULL;
char *iface = NULL;
char *optsrc, *optdst, *srceaddr, *dsteaddr;
static struct in6_addr ip6src, ip6dst;
struct in6_addr *optsrcn, *optdstn;
int sflag = 0, dflag = 0, nflag = 0;
static void usage __P((void));
int main __P((int, char *[]));
int bpf_open __P((char *));
static int linkhdrlen __P((int, char *));
static void form __P((int, char *));
static void form_ether __P((void));

static void
usage()
{
	fprintf(stderr,
		"usage: v6test [-d dstaddr] [-f configfile] [-i interface] "
		"[-n] [-s srcaddr] testname [testname...]\n");
	exit(1);
}

int
main(argc, argv)
	int argc;
	char *argv[];
{
	extern char *optarg;
	extern int optind;
	int size, ch;
	int fd = -1;
	int linkhdr;
	
	while ((ch = getopt(argc, argv, "d:f:i:ns:")) != -1)
		switch(ch) {
		case 's':
			if (inet_pton(AF_INET6, optarg, &ip6src) != 1) {
				optdst = optarg;
			} else
				optsrcn = &ip6src;
			break;
		case 'f':
			conffile = optarg;
			break;
		case 'd':
			if(inet_pton(AF_INET6, optarg, &ip6dst) != 1) {
				optdst = optarg;
			} else
				optdstn = &ip6dst;
			break;
		case 'i':
			iface = optarg;
			break;
		case 'n':
			nflag++;
			break;
		default:
			usage();
			/*NOTREACHED*/
		}
	argc -= optind;
	argv += optind;

	if (argc == 0) {
		usage();
		/*NOTREACHED*/
	}

	bzero(buf, sizeof(buf));
	if (nflag == 0)
		fd = bpf_open(iface);
	linkhdr = linkhdrlen(fd, iface);
	if (linkhdr < 0) {
		errx(1, "unsupported interface %s", iface);
		/*NOTREACHED*/
	}
	for (; argc > 0; argv++, argc--) {
		size = getconfig(*argv, buf + linkhdr);
		form(fd, iface);
		if (size && nflag == 0)
			write(fd, buf, size + linkhdr);
	}
	exit(0);
}

int
bpf_open(char *iface)
{
	/* based on dhcpc_subr.c(dhcp-1.3.9p2) */
	int n = 0, fd;
	char dev[16];
	struct ifreq ifr;
#ifdef HAVE_LIBPCAP
	char pcaperrbuf[PCAP_ERRBUF_SIZE];
#endif 
	
#define NBPFILTER 4
	do {
		sprintf(dev, "/dev/bpf%d", n++);
		fd = open(dev, O_RDWR);
	} while (fd < 0 && n < NBPFILTER);
	if (fd < 0) {
		fprintf(stderr, "Can't open bpf\n");
		exit(1);
	}

	if (ioctl(fd, BIOCIMMEDIATE, &n) < 0) {
		perror("ioctl(BIOCIMMEDIATE)");
		exit(1);
	}

	bzero(&ifr, sizeof(ifr));
#ifdef HAVE_LIBPCAP
	if (iface == NULL &&
	    (iface = pcap_lookupdev(pcaperrbuf)) == NULL) {
		fprintf(stderr, "pcap_lookupdev: %s\n", pcaperrbuf);
		exit(1);
	}
	fprintf(stderr, "outgoing interface is %s\n", iface);
#else
	if (iface == NULL) {
		fprintf(stderr, "interface must be specified\n");
		exit(1);
	}
#endif 
	strcpy(ifr.ifr_name, iface);
	if (ioctl(fd, BIOCSETIF, &ifr) < 0) {
		perror("ioctl(BIOCSETIF)");
		return(-1);
	}
	return(fd);
}

static int
linkhdrlen(fd, iface)
	int fd;
	char *iface;
{
	u_int v;

	if (fd < 0)
		return sizeof(struct ether_header);

	if (ioctl(fd, BIOCGDLT, (caddr_t)&v) < 0) {
		err(1, "ioctl(BIOCGDLT)");
		/*NOTREACHED*/
	}
	switch (v) {
	case DLT_EN10MB:
		return sizeof(struct ether_header);
	case DLT_NULL:
		return 0;
	}

	return -1;
}

static void
form(fd, iface)
	int fd;
	char *iface;
{
	u_int v;

	if (fd < 0)
		return;

	if (ioctl(fd, BIOCGDLT, (caddr_t)&v) < 0) {
		err(1, "ioctl(BIOCGDLT)");
		/*NOTREACHED*/
	}
	switch (v) {
	case DLT_EN10MB:
		form_ether();
		break;
	case DLT_NULL:
		break;
	}
}

static void
form_ether()
{
	struct ether_header *ether;
	struct ip6_hdr *ip;
	
	ether = (struct ether_header *)buf;
	ip = (struct ip6_hdr *)(ether + 1);

#ifndef ETHERTYPE_IPV6
#define	ETHERTYPE_IPV6	0x86dd /* Ether type for IPv6 */
#endif
	ether->ether_type = htons(ETHERTYPE_IPV6);
	if (srceaddr)
		bcopy(srceaddr, ether->ether_shost, 6);
	if (dsteaddr)
		bcopy(dsteaddr, ether->ether_dhost, 6);
	else if (IN6_IS_ADDR_MULTICAST(&(ip->ip6_dst))) {
		ether->ether_dhost[0] = 0x33;
		ether->ether_dhost[1] = 0x33;
		ether->ether_dhost[2] = ip->ip6_dst.s6_addr[12];
		ether->ether_dhost[3] = ip->ip6_dst.s6_addr[13];
		ether->ether_dhost[4] = ip->ip6_dst.s6_addr[14];
		ether->ether_dhost[5] = ip->ip6_dst.s6_addr[15];
	} else {
		ether->ether_dhost[0] = ip->ip6_dst.s6_addr[8] & 0xfd;
		ether->ether_dhost[1] = ip->ip6_dst.s6_addr[9];
		ether->ether_dhost[2] = ip->ip6_dst.s6_addr[10];
		ether->ether_dhost[3] = ip->ip6_dst.s6_addr[13];
		ether->ether_dhost[4] = ip->ip6_dst.s6_addr[14];
		ether->ether_dhost[5] = ip->ip6_dst.s6_addr[15];
	}		

	cksum6();
}
